### PR TITLE
refactor(build): Support electrical revisions explicitly

### PR DESCRIPTION
**Pull request by sfoster1**
As the development of the OT-3 goes on, we start having different versions of printed circuit boards. Sometimes, those different versions need to have different code that interacts with them - in fact, frequently they do, since behavioral changes are why you make "new versions" of PCBAs.

This PR lets us easily make a different firmware build and bootloader for each electrical PCBA revision. The output hex files contain the revision in them, making them easier to pack in side by side to deliver to the robot for update purposes, where the robot can always pick the right one based on the response from a given board.

To make this a little easier to deal with, especially to add new revisions, a lot of the stuff that was copy-pasted in individual projects is now defined in a big old macro with a lot of arguments and a lot of documentation. I recommend reading the macro doc at the top of the file for an overview of the options. Most application firmwares don't have to (and don't) touch most of the switches, but the bootloaders need to not build themselves, and the rear-panel needs a different set of bootloader handlers and so on, so the switches tend to be used somewhere.

Finally, in the process of updating CI to deal with all the new files, I realized it was a huge pain, so I used cmake install directives to provide a dist/ directory that's easy for CI to pull stuff out of.

## Review Requests

This is kind of a big one. You can go commit-by-commit, and quite honestly skip some once you get the point.

## Testing

- [x] Upload all the firmwares and flash all the images and make sure it all still talks and blinks and moves.

Closes RET-1317

### CODE REVIEW

File: common/core/revision.c.in [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include "common/core/version.h"
+
+static const struct revision _revision = {
+    .primary = '${PRIMARY_REVISION}',
```
Comment (Laura-Danielle) [inline common/core/revision.c.in:4]: Would be nice to either put a link to the confluence page that explains what the primary and secondary is, or just put a little blurb here.

**Review by Laura-Danielle — APPROVED** [review]
Love it! One itty-bitty comment request.

**Review by fsinapi — APPROVED** [review]
This looks rad. Pulled the branch and tested it and things seem to be building fine.

File: common/core/revision.c.in (hunk 1 above)
Comment (sfoster1) [reply common/core/revision.c.in:4]: I think I might put a blurb in the readme instead - I don't like having that sort of content in a generated file

**Review by vegano1 — APPROVED** [review]

**Pull request merged**